components/raftstore/src/store/util.rs

```diff
@@ -836,6 +837,155 @@ impl Display for MsgType<'_> {
     }
 }
 
+/// `RegionReadProgress` is used to keep track of the replica's `safe_ts`, the replica can handle a read
+/// request directly without requiring leader lease or read index iff `safe_ts` >= `read_ts` (the `read_ts`
+/// is usually stale i.e seconds ago).
+///
+/// `safe_ts` is updated by the `(apply index, safe ts)` item:
+/// ```
+/// if self.applied_index >= item.apply_index {
+///     self.safe_ts = max(self.safe_ts, item.safe_ts)
+/// }
+/// ```
+///
+/// For the leader, the `(apply index, safe ts)` item is publish by the `resolved-ts` worker periodically.
+/// For the followers, the item is sync periodically from the leader through the `CheckLeader` rpc.
+///
+/// The intend is to make the item's `safe ts` larger (more up to date) and `apply index` smaller (require less data)
+#[derive(Debug)]
+pub struct RegionReadProgress {
```

Comment:
If `safe_ts` and `applied_index` are wrapped into a struct, then this can be defined without lock.

```
pub struct StaleReadState {
    safe_ts: AtomicU64,
    applied_index: AtomicU64,
}

pub struct RegionReadProgress {
    state: Arc<StaleReadSTate>,
    applied_index: u64,
    safe_ts: u64,
    pending_items: VecDeque<(u64, u64)>,
}
```

Reply:
Besides `state`, other fields in `RegionReadProgress` could be updated by different threads

Reply:
How? Isn't it only be updated by checkLeader RPC response? The response should be handled in peer, which can't be processed by multiple threads at the same time.

Reply:
The response will be handled by the `StoreFsm` to avoid dispatch messages to all `PeerFsm` since there could a large amount of them, and it could be processed by different threads with the `PeerFsm` at the same time.

Also, currently, for the leader, `RegionReadProgress` will be updated at the `resolved ts` worker directly, we can let the worker send message to `raftstore` doing update instead, but it just bring the same problem.